### PR TITLE
[SPARK-8563] [MLlib] Fixed a bug so that IndexedRowMatrix.computeSVD().U.numCols = k

### PR DESCRIPTION
**Pull request by lee19**
I'm sorry that I made https://github.com/apache/spark/pull/6949 closed by mistake.
I pushed codes again.

And, I added a test code.

> There is a bug that `U.numCols() = self.nCols` in `IndexedRowMatrix.computeSVD()`
> It should have been `U.numCols() = k = svd.U.numCols()`
> 
> ```
> self = U * sigma * V.transpose
> (m x n) = (m x n) * (k x k) * (k x n) //ASIS
> -->
> (m x n) = (m x k) * (k x k) * (k x n) //TOBE
> ```

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/IndexedRowMatrix.scala [hunk 1]
```diff
@@ -146,7 +146,7 @@ class IndexedRowMatrix(
       val indexedRows = indices.zip(svd.U.rows).map { case (i, v) =>
         IndexedRow(i, v)
       }
-      new IndexedRowMatrix(indexedRows, nRows, nCols)
+      new IndexedRowMatrix(indexedRows, svd.U.numRows(), svd.U.numCols())
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/IndexedRowMatrix.scala]: You will need to write: ```svd.U.numCols().toInt```

File: mllib/src/test/scala/org/apache/spark/mllib/linalg/distributed/IndexedRowMatrixSuite.scala [hunk 2]
```diff
@@ -135,6 +135,17 @@ class IndexedRowMatrixSuite extends SparkFunSuite with MLlibTestSparkContext {
...
   }
 
+  test("validate matrix sizes of svd") {
+    val k = 2
+    val A = new IndexedRowMatrix(indexedRows)
+    val svd = A.computeSVD(k, computeU = true)
+    assert(svd.U.numRows() === m)
+    assert(svd.U.numCols() === k)
+    assert(svd.s.size === k)
+    assert(svd.V.numRows() === n)
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/mllib/linalg/distributed/IndexedRowMatrixSuite.scala]: This and the next line should not use parentheses for the methods.

**Comment by jkbradley** [comment]
@lee19 Thanks for the PR!  Those 2 issues should be quick to fix.  But please run the tests locally before updating to make sure the update works.

**Comment by srowen** [comment]
OK to test

**Comment by srowen** [comment]
LGTM

**Comment by lee19** [comment]
I'm very sorry for build errors. :flushed: I should have spent time on it at least.
And, I'm sorry again for the failure of unit test this time.
I can't run unit tests locally, yet, because 
`spark/mlib$ ../build/mvn test` is being stuck at LogisticRegressionClusterSuite because of OutOfMemoryError.
Let me tell you after I finish it.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/IndexedRowMatrix.scala [hunk 3]
```diff
@@ -146,7 +146,7 @@ class IndexedRowMatrix(
       val indexedRows = indices.zip(svd.U.rows).map { case (i, v) =>
         IndexedRow(i, v)
       }
-      new IndexedRowMatrix(indexedRows, nRows, nCols)
+      new IndexedRowMatrix(indexedRows, svd.U.numRows(), svd.U.numCols().toInt)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/IndexedRowMatrix.scala]: To avoid extra job, we can keep `nRows`.

**Comment by srowen** [comment]
@lee19 can you update this per the comment about nRows? I am not sure why your test failed since it looks correct, as does your change. Yes maybe you can run just that test locally to figure out what's going on.

**Comment by lee19** [comment]
I just confirmed `spark$ build/mvn -pl mllib -Pyarn -Phadoop.version=2.4 -DwildcardSuites=org.apache.spark.mllib.linalg.distributed.IndexedRowMatrixSuite test` passed locally.
I don't know, neither, why the test build 963 of SparkQA failed few days ago.
@mengxr Thanks for the suggestion!

**Comment by mengxr** [comment]
LGTM. Merged into master and all branches since 1.0. Thanks!!

**Pull request closed without merging**